designs/priorityqueue.md

```diff
@@ -0,0 +1,100 @@
+Priority Queue
+===================
+
+This document describes the motivation behind implementing a priority queue
+in controller-runtime and its design details.
+
+## Motivation
+
+1. Controllers reconcile all objects during startup to account for changes in
+   the reconciliation logic. Some controllers also periodically re-reconcile
+   everything to account for out of band changes they do not get notified for,
+   this is for example common for controllers managing cloud resources. In both
+   these cases, the reconciliation of new or changed objects gets delayed,
+   resulting in poor user experience. [Example][0]
+2. There may be application-specific reason why some events are more important
+   than others, [Example][1]
+
+## Proposed changes
+
+Implement a priority queue in controller-runtime that exposes the following
+interface:
+
+```go
+type PriorityQueue[T comparable] interface {
+    // AddWithOpts adds one or more items to the workqueue. Items
+    // in the workqueue are de-duplicated, so there will only ever
+    // be one entry for a given key.
+    // Adding an item that is already there may update its wait
+    // period to the lowest of existing and new wait period or
+    // its priority to the highest of existing and new priority.
+    AddWithOpts(o AddOpts, items ...T)
+
+    // GetWithPriority returns an item and its priority. It allows
+    // a controller to re-use the priority if it enqueues an item
+    // again.
+    GetWithPriority() (item T, priority int, shutdown bool)
+
+    // workqueue.TypedRateLimitingInterface is kept for backwards
+    // compatibility.
+    workqueue.TypedRateLimitingInterface[T]
+}
+
+type AddOpts struct {
+    // After is a duration after which the object will be available for
+    // reconciliation. If the object is already in the workqueue, the
+    // lowest of existing and new After period will be used.
+    After       time.Duration
+
+    // Ratelimited specifies if the ratelimiter should be used to
+    // determine a wait period. If the object is already in the
+    // workqueue, the lowest of existing and new wait period will be
+    // used.
+    RateLimited bool
+
+    // Priority specifies the priority of the object. Objects with higher
+    // priority are returned before objects with lower priority. If the
+    // object is already in the workqueue, the priority will be updated
+    // to the highest of existing and new priority.
+    Priority    int
+}
```

Comment:
I’m wondering if we need to consider starvation scenarios whether some item is never retrieved because there’s always something of higher priority in the queue. It might be fair to say this is “user error”, but is there a way to detect / avoid / recover?

Reply:
Interesting point. I don't think avoid or recover, but maybe detect. I'll think about it

Reply:
Queue depth is definitely good enough in general to detect that the queue cannot be drained. Isn't there also a metric for the longest time something is in the queue?

Independent of that we could also consider increasing priority after each resync, so it eventually reaches or maybe even exceeds (not sure) the priority of regular events

Reply:
Fine for me!